resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -744,9 +746,23 @@ object ApplicationMaster extends Logging {
   private val EXIT_SECURITY = 14
   private val EXIT_EXCEPTION_USER_CLASS = 15
   private val EXIT_EARLY = 16
+  private val EXIT_EXCEPTION_USER_APP = 17
 
   private var master: ApplicationMaster = _
 
+  def shouldUnregister(exitCode: Int): Boolean = {
+    // Only when ApplicationMaster's exit code is "EXIT_REPORTER_FAILURE" or "EXIT_EARLY" or
+    // "EXIT_UNCAUGHT_EXCEPTION" we should let RM to do reattempt. If the exception is from
+    // Spark internally, for example illegal state, executor failure, user code bug, we
+    // should not trigger another attempt.
+    exitCode == EXIT_SUCCESS ||
```

Comment:
Maybe we could make this more readable, just like `Set(...).find(_ == exitCode).isEmpty`.